// ng-lightning/esm2015/lib/accordion/accordion.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, Output, EventEmitter, ChangeDetectionStrategy, ElementRef, Renderer2, ContentChildren, QueryList } from '@angular/core';
import { NglAccordionSection } from './accordion-section';
import { isOptionSelected, addOptionToSelection } from '../util/util';
import { InputBoolean } from '../util/convert';
let NglAccordion = class NglAccordion {
    constructor(element, renderer) {
        this.activeNameChange = new EventEmitter();
        /**
         * Whether we allow multiple sections open at a time.
         */
        this.multiple = false;
        renderer.addClass(element.nativeElement, 'slds-accordion');
    }
    toggle(section) {
        const active = addOptionToSelection(section.name, this.activeName, this.multiple, true);
        this.activeNameChange.emit(active);
    }
    isActive(section) {
        return isOptionSelected(section.name, this.activeName, this.multiple);
    }
};
NglAccordion.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
__decorate([
    Input(),
    __metadata("design:type", Object)
], NglAccordion.prototype, "activeName", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], NglAccordion.prototype, "activeNameChange", void 0);
__decorate([
    Input(), InputBoolean(),
    __metadata("design:type", Object)
], NglAccordion.prototype, "multiple", void 0);
__decorate([
    ContentChildren(NglAccordionSection),
    __metadata("design:type", QueryList)
], NglAccordion.prototype, "sections", void 0);
NglAccordion = __decorate([
    Component({
        selector: 'ngl-accordion,[ngl-accordion]',
        changeDetection: ChangeDetectionStrategy.OnPush,
        template: "\n<li *ngFor=\"let section of sections\" nglAccordionItem [isActive]=\"isActive(section)\" [section]=\"section\" (toggle)=\"toggle(section)\"></li>"
    }),
    __metadata("design:paramtypes", [ElementRef, Renderer2])
], NglAccordion);
export { NglAccordion };
