// ng-lightning/esm5/lib/tooltips/trigger.js
import { __assign, __decorate, __metadata, __param, __read, __spread } from "tslib";
import { Directive, Input, ElementRef, TemplateRef, ViewContainerRef, OnDestroy, OnChanges, SimpleChanges, Output, EventEmitter, Renderer2, HostListener, Optional, Inject } from '@angular/core';
import { OverlayRef, Overlay, FlexibleConnectedPositionStrategy } from '@angular/cdk/overlay';
import { ComponentPortal } from '@angular/cdk/portal';
import { map, distinctUntilChanged } from 'rxjs/operators';
import { NglTooltip } from './tooltip';
import { POSITION_MAP, DEFAULT_TOOLTIP_POSITIONS, getPlacementName } from '../util/overlay-position';
import { uniqueId } from '../util/util';
import { InputBoolean } from '../util/convert';
import { NGL_TOOLTIP_CONFIG, NglTooltipConfig } from './config';
var NglTooltipTrigger = /** @class */ (function () {
    function NglTooltipTrigger(defaultConfig, element, renderer, viewContainerRef, overlay) {
        this.element = element;
        this.renderer = renderer;
        this.viewContainerRef = viewContainerRef;
        this.overlay = overlay;
        /**
         * Emit an event when actual tooltip is shown or hidden.
         */
        this.nglTooltipOpenChange = new EventEmitter();
        this.uid = uniqueId('tooltip');
        /** Names of properties that should be proxy to child component. */
        this.needProxyProperties = new Set([
            'template',
            'placement',
            'uid',
            'tooltipClass',
        ]);
        this.openDelay = 0;
        this.closeDelay = 0;
        this.toggleTimeout = null;
        this.overlayListeners = new Set();
        this.config = __assign(__assign({}, new NglTooltipConfig()), defaultConfig);
        this.openAuto = this.config.openAuto;
        this.interactive = this.config.interactive;
        this.delay = this.config.delay;
        this.renderer.setAttribute(this.element.nativeElement, 'aria-describedby', this.uid);
    }
    Object.defineProperty(NglTooltipTrigger.prototype, "placement", {
        get: function () {
            return this._placement || this.config.placement;
        },
        /**
         * Position relative to host element.
         */
        set: function (placement) {
            if (placement === this.placement) {
                return;
            }
            this._placement = placement;
            if (this.overlayRef) {
                this.updatePosition();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NglTooltipTrigger.prototype, "delay", {
        /**
         * Delay in milliseconds until it opens/closes.
         */
        set: function (_delay) {
            var _a;
            var delay = Array.isArray(_delay) ? _delay : [_delay, _delay];
            _a = __read(delay.map(Number), 2), this.openDelay = _a[0], this.closeDelay = _a[1];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NglTooltipTrigger.prototype, "nglOpen", {
        get: function () {
            return this._open;
        },
        /**
         * Whether the floating tooltip is visible.
         */
        set: function (open) {
            if (open === this.nglOpen) {
                return;
            }
            open ? this.create() : this.detach();
            this._open = open;
        },
        enumerable: true,
        configurable: true
    });
    NglTooltipTrigger.prototype.ngOnChanges = function (changes) {
        var _this = this;
        if (this.nglOpen) {
            this.updateProxies(changes);
            Promise.resolve().then(function () {
                if (_this.overlayRef) {
                    _this.overlayRef.updatePosition();
                }
            });
        }
    };
    NglTooltipTrigger.prototype.onMouseOver = function () {
        this.open();
    };
    NglTooltipTrigger.prototype.onMouseOut = function () {
        var _this = this;
        this.close();
        if (this.overlayRef && !this.overlayElement && this.interactive) {
            this.overlayElement = this.overlayRef.overlayElement;
            this.overlayListeners.add(this.renderer.listen(this.overlayElement, 'mouseenter', function () { return _this.open(); }));
            this.overlayListeners.add(this.renderer.listen(this.overlayElement, 'mouseleave', function () { return _this.close(); }));
        }
    };
    NglTooltipTrigger.prototype.ngOnDestroy = function () {
        this.detach();
        this.close(0);
    };
    // Expose open method
    NglTooltipTrigger.prototype.open = function (delay) {
        if (delay === void 0) { delay = this.openDelay; }
        this.handle(true, delay);
    };
    // Expose close method
    NglTooltipTrigger.prototype.close = function (delay) {
        if (delay === void 0) { delay = this.closeDelay; }
        this.handle(false, delay);
    };
    // Expose toggle method
    NglTooltipTrigger.prototype.toggle = function () {
        this.nglOpen ? this.close(0) : this.open(0);
    };
    NglTooltipTrigger.prototype.handle = function (open, delay) {
        var _this = this;
        if (this.toggleTimeout !== null) {
            clearTimeout(this.toggleTimeout);
            this.toggleTimeout = null;
        }
        if (open !== this.nglOpen) {
            if (delay > 0) {
                this.toggleTimeout = setTimeout(function () {
                    _this.toggleTimeout = null;
                    _this.emitOpen(open);
                }, delay);
            }
            else {
                this.emitOpen(open);
            }
        }
    };
    NglTooltipTrigger.prototype.emitOpen = function (open) {
        if (this.openAuto) {
            this.nglOpen = open;
        }
        this.nglTooltipOpenChange.emit(open);
    };
    NglTooltipTrigger.prototype.create = function () {
        var _this = this;
        if (this.nglOpen) {
            return;
        }
        this.detach();
        var overlayRef = this.createOverlay();
        this.portal = this.portal || new ComponentPortal(NglTooltip, this.viewContainerRef);
        this.tooltip = overlayRef.attach(this.portal).instance;
        this.needProxyProperties.forEach(function (property) { return _this.updateTooltip(property, _this[property]); });
    };
    /** Detaches the currently-attached tooltip. */
    NglTooltipTrigger.prototype.detach = function () {
        if (this.overlayRef) {
            this.overlayRef.detach();
            this.overlayRef.dispose();
            this.overlayRef = null;
        }
        // Clean up the event listeners
        this.overlayListeners.forEach(function (unlisten) { return unlisten(); });
        this.overlayListeners.clear();
        // Clear the overlay reference used for interactive mode
        if (this.interactive) {
            this.overlayElement = null;
        }
        if (this.positionChangesSubscription) {
            this.positionChangesSubscription.unsubscribe();
            this.positionChangesSubscription = null;
        }
        this.tooltip = null;
    };
    /** Create the overlay config and position strategy */
    NglTooltipTrigger.prototype.createOverlay = function () {
        var _this = this;
        if (this.overlayRef) {
            return this.overlayRef;
        }
        // Create connected position strategy that listens for scroll events to reposition.
        var strategy = this.overlay.position()
            .flexibleConnectedTo(this.element)
            .withFlexibleDimensions(false)
            .withViewportMargin(8)
            .withPush(false);
        this.positionChangesSubscription = strategy.positionChanges
            .pipe(map(function (change) { return getPlacementName(change, _this.placement); }), distinctUntilChanged())
            .subscribe(function (placement) {
            _this.updatePosition();
            _this.updateTooltip('placement', placement);
        });
        this.overlayRef = this.overlay.create({
            positionStrategy: strategy,
            scrollStrategy: this.overlay.scrollStrategies.reposition(),
        });
        this.updatePosition();
        return this.overlayRef;
    };
    /** Updates the position of the current tooltip. */
    NglTooltipTrigger.prototype.updatePosition = function () {
        var position = this.overlayRef.getConfig().positionStrategy;
        position.withPositions(__spread([
            POSITION_MAP[this.placement].position
        ], DEFAULT_TOOLTIP_POSITIONS));
    };
    NglTooltipTrigger.prototype.updateTooltip = function (key, value) {
        this.tooltip[key] = value;
    };
    /**
     * Set inputs of child components when this component's inputs change.
     */
    NglTooltipTrigger.prototype.updateProxies = function (changes) {
        var _this = this;
        Object.keys(changes)
            .filter(function (key) { return _this.needProxyProperties.has(key); })
            .forEach(function (key) { return _this.updateTooltip(key, _this[key]); });
    };
    NglTooltipTrigger.ctorParameters = function () { return [
        { type: NglTooltipConfig, decorators: [{ type: Optional }, { type: Inject, args: [NGL_TOOLTIP_CONFIG,] }] },
        { type: ElementRef },
        { type: Renderer2 },
        { type: ViewContainerRef },
        { type: Overlay }
    ]; };
    __decorate([
        Input('nglTooltip'),
        __metadata("design:type", Object)
    ], NglTooltipTrigger.prototype, "template", void 0);
    __decorate([
        Input('nglTooltipPlacement'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], NglTooltipTrigger.prototype, "placement", null);
    __decorate([
        Input('nglTooltipDelay'),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], NglTooltipTrigger.prototype, "delay", null);
    __decorate([
        Input('nglTooltipOpen'),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], NglTooltipTrigger.prototype, "nglOpen", null);
    __decorate([
        Input('nglTooltipOpenAuto'), InputBoolean(),
        __metadata("design:type", Boolean)
    ], NglTooltipTrigger.prototype, "openAuto", void 0);
    __decorate([
        Input('nglTooltipInteractive'), InputBoolean(),
        __metadata("design:type", Boolean)
    ], NglTooltipTrigger.prototype, "interactive", void 0);
    __decorate([
        Input('nglTooltipClass'),
        __metadata("design:type", Object)
    ], NglTooltipTrigger.prototype, "tooltipClass", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], NglTooltipTrigger.prototype, "nglTooltipOpenChange", void 0);
    __decorate([
        HostListener('mouseenter'),
        HostListener('focus'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], NglTooltipTrigger.prototype, "onMouseOver", null);
    __decorate([
        HostListener('mouseleave'),
        HostListener('blur'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], NglTooltipTrigger.prototype, "onMouseOut", null);
    NglTooltipTrigger = __decorate([
        Directive({
            selector: '[nglTooltip]',
            exportAs: 'nglTooltip',
        }),
        __param(0, Optional()), __param(0, Inject(NGL_TOOLTIP_CONFIG)),
        __metadata("design:paramtypes", [NglTooltipConfig,
            ElementRef,
            Renderer2,
            ViewContainerRef,
            Overlay])
    ], NglTooltipTrigger);
    return NglTooltipTrigger;
}());
export { NglTooltipTrigger };
