// ng-lightning/esm5/lib/datepickers/input/datepicker-input.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, ElementRef, Renderer2, HostListener, OnDestroy } from '@angular/core';
import { NglDatepickerInput } from './datepicker-input';
var NglDatepickerInputDirective = /** @class */ (function () {
    function NglDatepickerInputDirective(element, renderer, datepickerInput) {
        this.element = element;
        this.renderer = renderer;
        this.datepickerInput = datepickerInput;
        renderer.addClass(element.nativeElement, 'slds-input');
        renderer.setAttribute(element.nativeElement, 'autocomplete', 'off');
        renderer.setAttribute(element.nativeElement, 'id', this.datepickerInput.uid);
        this.datepickerInput.inputEl = this;
    }
    NglDatepickerInputDirective.prototype.onClick = function () {
        this.datepickerInput.onTriggerClick('input');
    };
    NglDatepickerInputDirective.prototype.onKeydown = function (evt) {
        this.datepickerInput.onKeyboardInput(evt);
    };
    NglDatepickerInputDirective.prototype.onInput = function () {
        var _this = this;
        setTimeout(function () { return _this.datepickerInput.onInputChange(); }, 0);
    };
    NglDatepickerInputDirective.prototype.onBlur = function () {
        this.datepickerInput.onBlur();
    };
    NglDatepickerInputDirective.prototype.setPlaceholder = function (placeholder) {
        this.renderer.setAttribute(this.element.nativeElement, 'placeholder', placeholder);
    };
    NglDatepickerInputDirective.prototype.setDisabled = function (disabled) {
        this.renderer.setProperty(this.element.nativeElement, 'disabled', disabled);
    };
    NglDatepickerInputDirective.prototype.ngOnDestroy = function () {
        this.datepickerInput.inputEl = null;
    };
    NglDatepickerInputDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: NglDatepickerInput }
    ]; };
    __decorate([
        HostListener('click'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], NglDatepickerInputDirective.prototype, "onClick", null);
    __decorate([
        HostListener('keydown', ['$event']),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], NglDatepickerInputDirective.prototype, "onKeydown", null);
    __decorate([
        HostListener('input'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], NglDatepickerInputDirective.prototype, "onInput", null);
    __decorate([
        HostListener('blur'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], NglDatepickerInputDirective.prototype, "onBlur", null);
    NglDatepickerInputDirective = __decorate([
        Directive({
            selector: 'input[nglDatepickerInput]',
            exportAs: 'nglDatepickerInput'
        }),
        __metadata("design:paramtypes", [ElementRef,
            Renderer2,
            NglDatepickerInput])
    ], NglDatepickerInputDirective);
    return NglDatepickerInputDirective;
}());
export { NglDatepickerInputDirective };
